# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/MediaTrackGraph.cpp
# Commit: 8cb7b8231c98
# Full Hash: 8cb7b8231c980aa7830fad2ecc34cb128417d511
# Author: Chun-Min Chang <chun.m.chang@gmail.com>
# Date: 2022-01-28 04:41:16
# Regressor Bug: 1742655
# File Overlap Count: 1
# Description:
#   Bug 1742655 - Open/Close audio input via NativeInputTrack r=pehrsons
#   
#   Move the audio-opening check logic and the creation of NativeInputTrack
#   in one place instead of seperating them in different functions. This
#   reduces the complexity of audio-opening logic in MediaTrackGraph and
# ==============================================================================

diff -r a1def130c105 -r 8cb7b8231c98 dom/media/MediaTrackGraph.cpp
--- a/dom/media/MediaTrackGraph.cpp	Thu Jan 27 22:29:34 2022 +0000
+++ b/dom/media/MediaTrackGraph.cpp	Thu Jan 27 22:29:35 2022 +0000
@@ -280,7 +280,7 @@
 
   // We may not have audio input device when we only have AudioNodeTracks. But
   // if audioTrackPresent is false, we must have no input device.
-  MOZ_DIAGNOSTIC_ASSERT_IF(!audioTrackPresent, !mNativeInputTrack);
+  MOZ_DIAGNOSTIC_ASSERT_IF(!audioTrackPresent, !mNativeInputTrackOnGraph);
 
   return audioTrackPresent;
 }
@@ -626,143 +626,79 @@
   return ticksWritten;
 }
 
-NativeInputTrack* MediaTrackGraphImpl::GetOrCreateDeviceTrack(
-    CubebUtils::AudioDeviceID aID, const PrincipalHandle& aPrincipalHandle) {
+NativeInputTrack* MediaTrackGraphImpl::GetNativeInputTrack() {
   MOZ_ASSERT(NS_IsMainThread());
-
-  RefPtr<NativeInputTrack>& t = mDeviceTracks.LookupOrInsertWith(aID, [&] {
-    NativeInputTrack* track =
-        NativeInputTrack::Create(this, aID, aPrincipalHandle);
-    LOG(LogLevel::Debug,
-        ("Create NativeInputTrack %p for device %p", track, aID));
-    return do_AddRef(track);
-  });
-  MOZ_DIAGNOSTIC_ASSERT(t->mPrincipalHandle == aPrincipalHandle,
-                        "Principal should match");
-
-  return t.get();
+  return mNativeInputTrackOnMain.get();
 }
 
-void MediaTrackGraphImpl::OpenAudioInputImpl(CubebUtils::AudioDeviceID aID,
-                                             AudioDataListener* aListener,
-                                             NativeInputTrack* aInputTrack) {
+void MediaTrackGraphImpl::OpenAudioInputImpl(NativeInputTrack* aTrack) {
   MOZ_ASSERT(OnGraphThread());
   LOG(LogLevel::Debug,
-      ("%p OpenAudioInputImpl: NativeInputTrack %p for device %p", this,
-       aInputTrack, aID));
-
-  if (mNativeInputTrack && mNativeInputTrack->mDeviceId != aID) {
-    // We don't support opening multiple input device in a graph for now.
-    LOG(LogLevel::Debug, ("%p Device %p is not native device. Cannot open %p!",
-                          this, aID, aInputTrack));
+      ("%p OpenAudioInputImpl: device %p", this, aTrack->mDeviceId));
+
+  if (mNativeInputTrackOnGraph) {
+    MOZ_ASSERT_UNREACHABLE(
+        "We cannot open device twice, and we don't support multiple inputs for "
+        "now");
     return;
   }
 
   LOG(LogLevel::Debug,
-      ("%p Device %p is native device. Open %p", this, aID, aInputTrack));
-
-  // Only allow one device per MTG (hence, per document), but allow opening a
-  // device multiple times
-  MOZ_ASSERT_IF(mNativeInputTrack, mNativeInputTrack == aInputTrack);
-  nsTArray<RefPtr<AudioDataListener>>& listeners = aInputTrack->mDataUsers;
-  MOZ_ASSERT(!listeners.Contains(aListener), "Don't add a listener twice.");
-  listeners.AppendElement(aListener);
-  if (listeners.Length() == 1) {  // first open for this device
-    mNativeInputTrack = aInputTrack;
-    mInputDeviceID = aID;
-    // Switch Drivers since we're adding input (to input-only or full-duplex)
-    AudioCallbackDriver* driver = new AudioCallbackDriver(
-        this, CurrentDriver(), mSampleRate, AudioOutputChannelCount(),
-        AudioInputChannelCount(), mOutputDeviceID, mInputDeviceID,
-        AudioInputDevicePreference());
-    LOG(LogLevel::Debug,
-        ("%p OpenAudioInput: starting new AudioCallbackDriver(input) %p", this,
-         driver));
-    SwitchAtNextIteration(driver);
-  } else {
-    ReevaluateInputDevice();
-  }
+      ("%p Open audio input on device %p", this, aTrack->mDeviceId));
+
+  mNativeInputTrackOnGraph = aTrack;
+  mInputDeviceID = aTrack->mDeviceId;
+  // Switch Drivers since we're adding input (to input-only or full-duplex)
+  AudioCallbackDriver* driver = new AudioCallbackDriver(
+      this, CurrentDriver(), mSampleRate, AudioOutputChannelCount(),
+      AudioInputChannelCount(), mOutputDeviceID, mInputDeviceID,
+      AudioInputDevicePreference());
+  LOG(LogLevel::Debug,
+      ("%p OpenAudioInput: starting new AudioCallbackDriver(input) %p", this,
+       driver));
+  SwitchAtNextIteration(driver);
 }
 
-nsresult MediaTrackGraphImpl::OpenAudioInput(CubebUtils::AudioDeviceID aID,
-                                             AudioDataListener* aListener) {
+void MediaTrackGraphImpl::OpenAudioInput(NativeInputTrack* aTrack) {
   MOZ_ASSERT(NS_IsMainThread());
+  MOZ_ASSERT(aTrack);
+
+  LOG(LogLevel::Debug, ("%p OpenInput: NativeInputTrack %p for device %p", this,
+                        aTrack, aTrack->mDeviceId));
+
   class Message : public ControlMessage {
    public:
-    Message(MediaTrackGraphImpl* aGraph, CubebUtils::AudioDeviceID aID,
-            AudioDataListener* aListener, NativeInputTrack* aInputTrack)
-        : ControlMessage(nullptr),
-          mGraph(aGraph),
-          mID(aID),
-          mListener(aListener),
-          mInputTrack(aInputTrack) {}
+    Message(MediaTrackGraphImpl* aGraph, NativeInputTrack* aInputTrack)
+        : ControlMessage(nullptr), mGraph(aGraph), mInputTrack(aInputTrack) {}
     void Run() override {
       TRACE("MTG::OpenAudioInputImpl ControlMessage");
-      mGraph->OpenAudioInputImpl(mID, mListener, mInputTrack);
+      mGraph->OpenAudioInputImpl(mInputTrack);
     }
     MediaTrackGraphImpl* mGraph;
-    CubebUtils::AudioDeviceID mID;
-    RefPtr<AudioDataListener> mListener;
     NativeInputTrack* mInputTrack;
   };
 
-  auto result = mDeviceTracks.Lookup(aID);
-  MOZ_ASSERT(result);
-  MOZ_ASSERT(result.Data());
-  size_t users = result.Data()->AddUser();
-
-  LOG(LogLevel::Debug,
-      ("%p OpenInput: NativeInputTrack %p for device %p has %zu users now",
-       this, result.Data().get(), aID, users));
+  MOZ_ASSERT(!mNativeInputTrackOnMain);
+  mNativeInputTrackOnMain = aTrack;
 
   // XXX Check not destroyed!
-  this->AppendMessage(
-      MakeUnique<Message>(this, aID, aListener, result.Data().get()));
-  return NS_OK;
+  this->AppendMessage(MakeUnique<Message>(this, aTrack));
 }
 
-void MediaTrackGraphImpl::CloseAudioInputImpl(CubebUtils::AudioDeviceID aID,
-                                              AudioDataListener* aListener,
-                                              NativeInputTrack* aInputTrack) {
+void MediaTrackGraphImpl::CloseAudioInputImpl(CubebUtils::AudioDeviceID aID) {
   MOZ_ASSERT(OnGraphThread());
 
-  LOG(LogLevel::Debug,
-      ("%p CloseAudioInputImpl: NativeInputTrack %p for device %p", this,
-       aInputTrack, aID));
-
-  if (!mNativeInputTrack || mNativeInputTrack->mDeviceId != aID) {
-    LOG(LogLevel::Debug,
-        ("%p Device %p is not native device. Do nothing for %p", this, aID,
-         aInputTrack));
+  LOG(LogLevel::Debug, ("%p CloseAudioInputImpl: device %p", this, aID));
+
+  if (!mNativeInputTrackOnGraph || mNativeInputTrackOnGraph->mDeviceId != aID) {
+    LOG(LogLevel::Debug, ("%p Device %p is not native device", this, aID));
     return;
   }
 
-  LOG(LogLevel::Debug,
-      ("%p Device %p is native device. Close %p", this, aID, aInputTrack));
-
-  MOZ_ASSERT(mNativeInputTrack == aInputTrack);
-  nsTArray<RefPtr<AudioDataListener>>& listeners = aInputTrack->mDataUsers;
-  DebugOnly<bool> wasPresent = listeners.RemoveElement(aListener);
-  MOZ_ASSERT(wasPresent);
-
-  // Breaks the cycle between the MTG and the listener.
-  aListener->Disconnect(this);
-
-  if (!listeners.IsEmpty()) {
-    LOG(LogLevel::Debug,
-        ("%p NativeInputTrack %p for device %p still has consumer", this,
-         aInputTrack, aID));
-    ReevaluateInputDevice();
-    return;
-  }
-
-  LOG(LogLevel::Debug,
-      ("%p NativeInputTrack %p for device %p has no consumer now", this,
-       aInputTrack, aID));
-
-  mNativeInputTrack = nullptr;  // reset to default
-  mInputDeviceID = nullptr;  // reset to default
-
+  LOG(LogLevel::Debug, ("%p Close device %p", this, aID));
+
+  mNativeInputTrackOnGraph = nullptr;  // reset to default
+  mInputDeviceID = nullptr;            // reset to default
 
   // Switch Drivers since we're adding or removing an input (to nothing/system
   // or output only)
@@ -823,71 +759,34 @@
       });
 }
 
-void MediaTrackGraphImpl::CloseAudioInput(CubebUtils::AudioDeviceID aID,
-                                          AudioDataListener* aListener) {
+void MediaTrackGraphImpl::CloseAudioInput(NativeInputTrack* aTrack) {
   MOZ_ASSERT(NS_IsMainThread());
   class Message : public ControlMessage {
    public:
-    Message(MediaTrackGraphImpl* aGraph, CubebUtils::AudioDeviceID aID,
-            AudioDataListener* aListener, NativeInputTrack* aInputTrack)
-        : ControlMessage(nullptr),
-          mGraph(aGraph),
-          mID(aID),
-          mListener(aListener),
-          mInputTrack(aInputTrack) {}
+    Message(MediaTrackGraphImpl* aGraph, CubebUtils::AudioDeviceID aID)
+        : ControlMessage(nullptr), mGraph(aGraph), mID(aID) {}
     void Run() override {
       TRACE("MTG::CloseAudioInputImpl ControlMessage");
-      mGraph->CloseAudioInputImpl(mID, mListener, mInputTrack);
+      mGraph->CloseAudioInputImpl(mID);
     }
     MediaTrackGraphImpl* mGraph;
     CubebUtils::AudioDeviceID mID;
-    RefPtr<AudioDataListener> mListener;
-    NativeInputTrack* mInputTrack;
   };
 
-  auto result = mDeviceTracks.Lookup(aID);
-  MOZ_ASSERT(result);
-  MOZ_ASSERT(result.Data());
-  size_t users = result.Data()->RemoveUser();
-
-  LOG(LogLevel::Debug,
-      ("%p: CloseInput: NativeInputTrack %p for device %p has %zu users now",
-       this, result.Data().get(), aID, users));
-
-  this->AppendMessage(
-      MakeUnique<Message>(this, aID, aListener, result.Data().get()));
-
-  // Remove the NativeInputTrack from mDeviceTracks if no AudioInputTrack needs
-  // it, so NativeInputTrack::Create can create a new NativeInputTrack when it's
-  // called for the same aID. The NativeInputTrack will still be alive after
-  // it's removed from mDeviceTracks since AddTrack called via
-  // NativeInputTrack::Create will call NS_ADDREF to it and it will be alive
-  // until its NS_RELEASE is called via NativeInputTrack::DestroyImpl(). Note
-  // that NativeInputTrack::Destroy() must be called after the above message is
-  // appended so NativeInputTrack::DestroyImpl() will be run after
-  // CloseAudioInputImpl(). Therefore, the NativeInputTrack will be alive when
-  // running CloseAudioInputImpl()
-  if (users == 0) {
-    LOG(LogLevel::Debug,
-        ("%p: CloseInput: NativeInputTrack %p for device %p is removed from "
-         "mDeviceTracks",
-         this, result.Data().get(), aID));
-
-    result.Data()->Destroy();
-    bool r = mDeviceTracks.Remove(aID);
-    MOZ_ASSERT(r);
-    Unused << r;
-  }
+  // NativeInputTrack is still alive (in mTracks) even we remove it here.
+  mNativeInputTrackOnMain = nullptr;
+
+  this->AppendMessage(MakeUnique<Message>(this, aTrack->mDeviceId));
 }
 
 // All AudioInput listeners get the same speaker data (at least for now).
 void MediaTrackGraphImpl::NotifyOutputData(AudioDataValue* aBuffer,
                                            size_t aFrames, TrackRate aRate,
                                            uint32_t aChannels) {
-  if (!mNativeInputTrack) {
+  if (!mNativeInputTrackOnGraph) {
     return;
   }
-  MOZ_ASSERT(mNativeInputTrack->mDeviceId == mInputDeviceID);
+  MOZ_ASSERT(mNativeInputTrackOnGraph->mDeviceId == mInputDeviceID);
 
   for (const auto& track : mTracks) {
     if (const auto& t = track->AsAudioInputTrack(); t) {
@@ -897,11 +796,11 @@
 }
 
 void MediaTrackGraphImpl::NotifyInputStopped() {
-  if (!mNativeInputTrack) {
+  if (!mNativeInputTrackOnGraph) {
     return;
   }
-  MOZ_ASSERT(mNativeInputTrack->mDeviceId == mInputDeviceID);
-  mNativeInputTrack->NotifyInputStopped(this);
+  MOZ_ASSERT(mNativeInputTrackOnGraph->mDeviceId == mInputDeviceID);
+  mNativeInputTrackOnGraph->NotifyInputStopped(this);
 }
 
 void MediaTrackGraphImpl::NotifyInputData(const AudioDataValue* aBuffer,
@@ -911,22 +810,22 @@
   // Either we have an audio input device, or we just removed the audio input
   // this iteration, and we're switching back to an output-only driver next
   // iteration.
-  MOZ_ASSERT(mNativeInputTrack || Switching());
-  if (!mNativeInputTrack) {
+  MOZ_ASSERT(mNativeInputTrackOnGraph || Switching());
+  if (!mNativeInputTrackOnGraph) {
     return;
   }
-  MOZ_ASSERT(mNativeInputTrack->mDeviceId == mInputDeviceID);
-  mNativeInputTrack->NotifyInputData(this, aBuffer, aFrames, aRate, aChannels,
-                                     aAlreadyBuffered);
+  MOZ_ASSERT(mNativeInputTrackOnGraph->mDeviceId == mInputDeviceID);
+  mNativeInputTrackOnGraph->NotifyInputData(this, aBuffer, aFrames, aRate,
+                                            aChannels, aAlreadyBuffered);
 }
 
 void MediaTrackGraphImpl::DeviceChangedImpl() {
   MOZ_ASSERT(OnGraphThread());
-  if (!mNativeInputTrack) {
+  if (!mNativeInputTrackOnGraph) {
     return;
   }
-  MOZ_ASSERT(mNativeInputTrack->mDeviceId == mInputDeviceID);
-  mNativeInputTrack->DeviceChanged(this);
+  MOZ_ASSERT(mNativeInputTrackOnGraph->mDeviceId == mInputDeviceID);
+  mNativeInputTrackOnGraph->DeviceChanged(this);
 }
 
 void MediaTrackGraphImpl::SetMaxOutputChannelCount(uint32_t aMaxChannelCount) {
@@ -3949,22 +3848,22 @@
 uint32_t MediaTrackGraphImpl::AudioInputChannelCount() {
   MOZ_ASSERT(OnGraphThreadOrNotRunning());
 
-  if (!mNativeInputTrack) {
+  if (!mNativeInputTrackOnGraph) {
     return 0;
   }
-  MOZ_ASSERT(mNativeInputTrack->mDeviceId == mInputDeviceID);
-  return mNativeInputTrack->MaxRequestedInputChannels();
+  MOZ_ASSERT(mNativeInputTrackOnGraph->mDeviceId == mInputDeviceID);
+  return mNativeInputTrackOnGraph->MaxRequestedInputChannels();
 }
 
 AudioInputType MediaTrackGraphImpl::AudioInputDevicePreference() {
   MOZ_ASSERT(OnGraphThreadOrNotRunning());
 
-  if (!mNativeInputTrack) {
+  if (!mNativeInputTrackOnGraph) {
     return AudioInputType::Unknown;
   }
-  MOZ_ASSERT(mNativeInputTrack->mDeviceId == mInputDeviceID);
-  return mNativeInputTrack->HasVoiceInput() ? AudioInputType::Voice
-                                            : AudioInputType::Unknown;
+  MOZ_ASSERT(mNativeInputTrackOnGraph->mDeviceId == mInputDeviceID);
+  return mNativeInputTrackOnGraph->HasVoiceInput() ? AudioInputType::Voice
+                                                   : AudioInputType::Unknown;
 }
 
 // nsIThreadObserver methods